### PR TITLE
Generate serialisers with sjson-new

### PR DESCRIPTION
**Pull request by Duhemm**

### CODE REVIEW

File: library/src/test/scala/SchemaExample.scala [hunk 1]
```diff
@@ -632,6 +632,138 @@ object NewSchema {
           |    }
           |}""".stripMargin)
 
+  val completeExampleCodeSerializer =
+    """package com.example
+      |import _root_.sjsonnew._
+      |import _root_.sjsonnew.BasicJsonProtocol._
```
Comment (eed3si9n) [inline library/src/test/scala/SchemaExample.scala]: BasicJsonProtocol is useful only for demo purpose.
I think what you should generate is a trait with a self type annotation of any necessary parent JSON protocols.

File: plugin/src/main/scala/Plugin.scala [hunk 2]
```diff
@@ -10,10 +10,16 @@ object DatatypePlugin extends AutoPlugin {
 
   object autoImport {
     val generateDatatypes = taskKey[Seq[File]]("Generate datatypes.")
+    val createDatatypes = settingKey[Boolean]("Whether to generate the datatypes or not.")
+    val createJsonFormats = settingKey[Boolean]("Whether to generate the JsonFormats for the datatypes or not.")
     val datatypeJavaLazy = settingKey[String]("Interface to use to provide laziness in Java.")
     val datatypeSource = settingKey[File]("Datatype source directory.")
     val datatypeScalaFileNames = settingKey[Definition => File]("Mapping from `Definition` to file for Scala generator.")
     val datatypeScalaSealProtocols = settingKey[Boolean]("Seal abstract classes representing `Protocol`s in Scala.")
+    val datatypeSerializerPackage = settingKey[Option[String]]("Package in which to put the serializer object.")
```
Comment (eed3si9n) [inline plugin/src/main/scala/Plugin.scala]: I think we should avoid the term "serializer" and call it "Json protocol" or "codec".

File: plugin/src/main/scala/Plugin.scala [hunk 3]
```diff
@@ -22,19 +28,31 @@ object DatatypePlugin extends AutoPlugin {
...
       datatypeSource in generateDatatypes := Defaults.configSrcSub(sourceDirectory).value / "datatype",
       sourceManaged in generateDatatypes := sourceManaged.value,
       datatypeScalaFileNames in generateDatatypes := scalaDef2File,
       // We cannot enable this by default, because the default function for naming Scala files that we provide
       // will create a separate file for every `Definition`.
       datatypeScalaSealProtocols in generateDatatypes := false,
+      datatypeSerializerPackage in generateDatatypes := Some("serialization"),
+      datatypeSerializerName in generateDatatypes := "Serializer",
+      datatypeSerializerParents in generateDatatypes := Nil,
+      datatypeInstantiateJavaLazy in generateDatatypes := { (e: String) => s"xsbti.SafeLazy($e)" },
```
Comment (eed3si9n) [inline plugin/src/main/scala/Plugin.scala]: sbt-datatype should generate this stuff, for all non-sbt usages (if any).

File: plugin/src/sbt-test/sbt-datatype/mixed-schema/build.sbt [hunk 4]
```diff
@@ -1 +1,5 @@
 name := "example"
+
+datatypeSerializerParents in generateDatatypes in Compile := Seq("com.example.serialization.FileFormat")
+
+libraryDependencies += "com.eed3si9n" %% "sjson-new-core" % "0.4.0"
```
Comment (eed3si9n) [inline plugin/src/sbt-test/sbt-datatype/mixed-schema/build.sbt]: Could you test if you can roundtrip with the data using sjson-new-scalajson?

**Comment by eed3si9n** [comment]
Happy to see advancements here.

File: library/src/test/scala/SchemaExample.scala (hunk 1 above)
Comment (Duhemm) [inline library/src/test/scala/SchemaExample.scala]: That's how its done now.

For instance, for a simple schema that just defines an enumeration, we generate: 

```scala
import _root_.sjsonnew.{ deserializationError, serializationError, Builder, JsonFormat, Unbuilder }
trait simpleEnumerationExampleFormat { self: sbt.datatype.StringFormat =>
  implicit lazy val simpleEnumerationExampleFormat: JsonFormat[simpleEnumerationExample] = new JsonFormat[simpleEnumerationExample] {
    override def read[J](jsOpt: Option[J], unbuilder: Unbuilder[J]): simpleEnumerationExample = {
      jsOpt match {
        case Some(js) =>
          unbuilder.readString(js) match {
            case "first" => simpleEnumerationExample.first
            case "second" => simpleEnumerationExample.second
          }
        case None =>
          deserializationError("Expected JsString but found None")
      }
    }
    override def write[J](obj: simpleEnumerationExample, builder: Builder[J]): Unit = {
      val str = obj match {
        case simpleEnumerationExample.first => "first"
        case simpleEnumerationExample.second => "second"
      }
      builder.writeString(str)
    }
  }
}

trait Codec { self: _root_.simpleEnumerationExampleFormat with sbt.datatype.StringFormat => }
object Codec extends Codec with _root_.simpleEnumerationExampleFormat with sbt.datatype.StringFormat
```

File: plugin/src/main/scala/Plugin.scala (hunk 2 above)
Comment (Duhemm) [inline plugin/src/main/scala/Plugin.scala]: Done

File: plugin/src/main/scala/Plugin.scala (hunk 3 above)
Comment (Duhemm) [inline plugin/src/main/scala/Plugin.scala]: Do you mean that sbt-datatype should provide a library that would define concrete instances of `xsbti.api.Lazy[T]`? At the moment there's a `settingKey` to define the name of the lazy type to use.

File: plugin/src/sbt-test/sbt-datatype/mixed-schema/build.sbt (hunk 4 above)
Comment (Duhemm) [inline plugin/src/sbt-test/sbt-datatype/mixed-schema/build.sbt]: See https://github.com/sbt/sbt-datatype/pull/15/commits/7a437435c4dedf0b714491a346648ecb3482cfbc#diff-eb80b92565fb84b97973f1ad4a942d29R1

**Pull request merged**